### PR TITLE
Issue 65

### PR DESCRIPTION
**Pull request by intendo**
This should update the documentation for the user to run apbs_tester on the examples.
This should also install the examples and missing test files.

### CODE REVIEW

**Comment by intendo** [comment]
This should close #65 .
@sobolevnrm can you download the Windows https://ci.appveyor.com/api/buildjobs/jt7g5brnbcl2c0tc/artifacts/build%2FAPBS-3.1.4.Windows.zip and double check that the testing documentation is correct and works?

**Comment by sobolevnrm** [comment]
> This should close #65 .
> @sobolevnrm can you download the Windows https://ci.appveyor.com/api/buildjobs/jt7g5brnbcl2c0tc/artifacts/build%2FAPBS-3.1.4.Windows.zip and double check that the testing documentation is correct and works?

Should I be using the RTD documentation for installation?  I don't think those included documentation on how to test... what am I missing?

**Comment by intendo** [comment]
Yes, you would use the new RTD (assuming they get built on pull requests).

The testing docs should be in docs/using/tests.html

**Comment by sobolevnrm** [comment]
> Yes, you would use the new RTD (assuming they get built on pull requests).
> 
> The testing docs should be in docs/using/tests.html

Got it.  They don't get built automatically but I can trigger a build manually.  I'll do that now.  Thanks!

**Comment by sobolevnrm** [comment]
The documentation does not render correctly; perhaps you were using Markdown instead of RestructuredText?  I'll try to fix it.

**Comment by sobolevnrm** [comment]
Meanwhile, in `share\apbs\tests` I ran `python apbs_tester.py -e ..\..\..\bin\apbs.exe -c test_cases.cfg` and got the following error:

```
BINARY_NAME:..\..\..\bin\apbs.exe
Testing all sections
The following sections will be tested: born, actin-dimer-auto, actin-dimer-parallel, alkanes, FKBP, hca-bind, ionize, ion-pmf, ion-protein, pka-lig, point-pmf, solv, geoflow, bem, bem-pKa, bem-binding-energy, pbam, pbsam, protein-rna   
================================================================================
Running tests for born section
--------------------------------------------------------------------------------
Testing forces from apbs-forces.in

Checking forces for input file apbs-forces.inChecking Polar ForcesChecking Apolar ForcesElapsed time: 0.002 seconds
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
Testing input file apbs-mol-auto.in

BINARY: ..\..\..\bin\apbs.exe
INPUT:  apbs-mol-auto.in
Traceback (most recent call last):
  File "C:\Users\bake113\OneDrive - PNNL\Desktop\APBS-3.1.4.Windows\share\apbs\tests\apbs_tester.py", line 431, in <module>
    sys.exit(main())
  File "C:\Users\bake113\OneDrive - PNNL\Desktop\APBS-3.1.4.Windows\share\apbs\tests\apbs_tester.py", line 416, in main
    run_test(
  File "C:\Users\bake113\OneDrive - PNNL\Desktop\APBS-3.1.4.Windows\share\apbs\tests\apbs_tester.py", line 230, in run_test
    computed_results = process_serial(binary, input_file)
  File "C:\Users\bake113\OneDrive - PNNL\Desktop\APBS-3.1.4.Windows\share\apbs\tests\apbs_tester.py", line 105, in process_serial
    with subprocess.Popen(
  File "C:\Users\bake113\Anaconda3\envs\foo\lib\subprocess.py", line 947, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Users\bake113\Anaconda3\envs\foo\lib\subprocess.py", line 1416, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
FileNotFoundError: [WinError 2] The system cannot find the file specified
```

**Comment by intendo** [comment]
My guess is the error is due to Python not being able to deal with spaces in path names on Windows.
I added some code to handle spaces in paths that should work.

**Comment by sobolevnrm** [comment]
> My guess is the error is due to Python not being able to deal with spaces in path names on Windows.
> I added some code to handle spaces in paths that should work.

Is there an updated ZIP file I can test?

**Comment by intendo** [comment]
Yes, https://ci.appveyor.com/api/buildjobs/dfqxhysdhmo7k9ty/artifacts/build%2FAPBS-3.1.4.Windows.zip

**Comment by sobolevnrm** [comment]
> Yes, https://ci.appveyor.com/api/buildjobs/dfqxhysdhmo7k9ty/artifacts/build%2FAPBS-3.1.4.Windows.zip

That gives the same error with a relative path -- regardless whether the location is in a path with spaces or not.  It seems to work when I specify the absolute path for a path with no spaces; it does not work when I specify an absolute path with spaces.

**Comment by intendo** [comment]
@sobolevnrm can you try to run 
     `cd share\apbs\tests` 
and  
     `python apbs_tester.py -e "C:\Users\bake113\OneDrive - PNNL\Desktop\APBS-3.1.4.Windows\share\apbs\tests\apbs_tester.py" -c test_cases.cfg`

I want to see if you have to use double quotes around the executable on Windows.

**Review by sobolevnrm — COMMENTED** [review]
> @sobolevnrm can you try to run
> `cd share\apbs\tests`
> and
> `python apbs_tester.py -e "C:\Users\bake113\OneDrive - PNNL\Desktop\APBS-3.1.4.Windows\share\apbs\tests\apbs_tester.py" -c test_cases.cfg`
> 
> I want to see if you have to use double quotes around the executable on Windows.

@intendo In this branch?

File: tests/apbs_tester.py [hunk 1]
```diff
@@ -24,12 +24,14 @@
 sys.path.insert(0, "../tools/manip")
 from inputgen import split_input  # noqa E402
 
+SEARCH_PATH = os.environ.get("PATH").split(os.pathsep)
```
[the remark this answers is not in the document]
Comment (intendo) [reply tests/apbs_tester.py:27]: I could not find a way to use `pathlib` to iterate over all the directories in the user's `PATH`. I use `pathlib` in the loop that iterates over `SEARCH_PATH`.

[the remark this answers is not in the document]
Comment (intendo) [reply tests/apbs_tester.py:27]: If you can find a way to use `pathlib` without using `os`, I am totally willing to use it.

**Comment by intendo** [comment]
@sobolevnrm you can use the last download or you can get last night's build at https://ci.appveyor.com/api/buildjobs/dx9fmde4m7wj7k3o/artifacts/build%2FAPBS-3.1.4.Windows.zip

**Comment by sobolevnrm** [comment]
> @sobolevnrm you can use the last download or you can get last night's build at https://ci.appveyor.com/api/buildjobs/dx9fmde4m7wj7k3o/artifacts/build%2FAPBS-3.1.4.Windows.zip
> 
> I added a test during the Windows build in Appveyor to do the following and it worked:
> 
> ```
> python apbs_tester.py -e "C:\Program Files\APBS\bin\apbs.exe" -c test_cases.cfg -t pbam
> ```
> 
> I just wanted you to verify that it works so we can document it and finish this merge.

Here's what I get:

```
python apbs_tester.py -e "C:\Users\bake113\OneDrive - PNNL\Desktop\APBS-3.1.4.Windows\share\apbs\tests\apbs_tester.py" -c test_cases.cfg
TESTING WITH BINARY_NAME:C:\Users\bake113\OneDrive - PNNL\Desktop\APBS-3.1.4.Windows\share\apbs\tests\apbs_tester.py
NOTE: Using apbs binary:C:\Users\bake113\OneDrive - PNNL\Desktop\APBS-3.1.4.Windows\share\apbs\tests\apbs_tester.py

Exception:[WinError 193] %1 is not a valid Win32 application
Testing all sections
The following sections will be tested: born, actin-dimer-auto, actin-dimer-parallel, alkanes, FKBP, hca-bind, ionize, ion-pmf, ion-protein, pka-lig, point-pmf, solv, geoflow, bem, bem-pKa, bem-binding-energy, pbam, pbsam, protein-rna
================================================================================
Running tests for born section
--------------------------------------------------------------------------------
Testing forces from apbs-forces.in

Checking forces for input file apbs-forces.inChecking Polar ForcesChecking Apolar ForcesElapsed time: 0.002 seconds
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
Testing input file apbs-mol-auto.in

BINARY:
INPUT:   apbs-mol-auto.in
COMMAND: ['', 'apbs-mol-auto.in']
Traceback (most recent call last):
  File "apbs_tester.py", line 438, in <module>
    sys.exit(main())
  File "apbs_tester.py", line 423, in main
    run_test(
  File "apbs_tester.py", line 242, in run_test
    computed_results = process_serial(binary, input_file)
  File "apbs_tester.py", line 116, in process_serial
    with subprocess.Popen(
  File "C:\Users\bake113\AppData\Local\Continuum\anaconda3\envs\osmolytes\lib\subprocess.py", line 854, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Users\bake113\AppData\Local\Continuum\anaconda3\envs\osmolytes\lib\subprocess.py", line 1307, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
OSError: [WinError 87] The parameter is incorrect
```

**Comment by intendo** [comment]
@sobolevnrm I cannot replicate this problem on a Windows box or VM.
I think the problem may be with your One Drive configuration (either being a network drive or some organizational permissions).
Can you install [APBS-3.1.4](https://ci.appveyor.com/api/buildjobs/dx9fmde4m7wj7k3o/artifacts/build%2FAPBS-3.1.4.Windows.zip) on a local drive?
You could even create a directory with spaces and hyphens to make sure that it works or not.
If it works locally, we may need to add a caveat in the documentation that installing on a Network drive may not work.

**Comment by sobolevnrm** [comment]
> @sobolevnrm I cannot replicate this problem on a Windows box or VM.
> I think the problem may be with your One Drive configuration (either being a network drive or some organizational permissions).
> Can you install [APBS-3.1.4](https://ci.appveyor.com/api/buildjobs/dx9fmde4m7wj7k3o/artifacts/build%2FAPBS-3.1.4.Windows.zip) on a local drive?
> You could even create a directory with spaces and hyphens to make sure that it works or not.
> If it works locally, we may need to add a caveat in the documentation that installing on a Network drive may not work.

That seems to work.  I'm not sure what's different but it is working now (from my OneDrive folder with spaces in the name, etc.)

**Comment by intendo** [comment]
closes #65

**Pull request merged**